p2p/starknet/iterator.go

```diff
@@ -0,0 +1,66 @@
+package starknet
+
+import (
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+)
+
+type iterator struct {
+	bcReader blockchain.Reader
+
+	blockNumber uint64
+	step        uint64
+	limit       uint64
+	forward     bool
+
+	err error
+}
+
+func newIterator(bcReader blockchain.Reader, blockNumber, limit, step uint64, forward bool) (*iterator, error) {
+	if step == 0 {
+		return nil, fmt.Errorf("step is zero")
+	}
+	if limit == 0 {
+		return nil, fmt.Errorf("limit is zero")
+	}
+
+	return &iterator{
+		bcReader:    bcReader,
+		blockNumber: blockNumber,
+		limit:       limit,
+		step:        step,
+		forward:     forward,
+	}, nil
+}
+
+func (it *iterator) Err() error {
+	return it.err
+}
+
+func (it *iterator) Valid() bool {
+	if it.limit <= 0 || it.err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (it *iterator) Next() bool {
+	if it.forward {
+		it.blockNumber += it.step
+	} else {
+		it.blockNumber -= it.step
+	}
+	it.limit--
+
+	return it.Valid()
+}
+
+func (it *iterator) Block() *core.Block {
```

Comment:
Would it be better to return the error directly, rather than requiring the caller to query `it.Err()`?

Reply:
Yeah, I was thinking about this for a while. For some time it was a common pattern for iterators to have `Next()`, `HasNext()`, and `Err` (see https://pkg.go.dev/bufio#Scanner for example). I'm not sure about it now, will reconsider it